applications/Preprocessor/lib/MergePlan.h

```diff
@@ -0,0 +1,474 @@
+    }
+
+    // Zero dimension MeshEntities 'nodes' are special in that each coordinate
+    // corresponds to a single node. Hence, merging coordinates is equivalent to
+    // merging the associated nodes. Which is much less complicated than for
+    // higher dimensional MeshEntities.
+    {
+        std::vector<MergePair> nodePairs;
+        nodePairs.reserve(pairing.size());
+        const auto& coordinates = mesh->getNodeCoordinates();
+        for (auto const& pair : pairing) {
+            nodePairs.emplace_back(coordinates[pair.first.id].nodeIndex,
+                                   coordinates[pair.second.id].nodeIndex);
+        }
+        merges[0] = groupMerges(nodePairs);
+    }
+
+    // For merging higher dimensional MeshEntities we need to know the elements
+    // that are adjacent to at least one of the coordinates for both the left
+    // and right coordinates.
+    MergeElements elements(mesh, pairing);
+    logger(INFO, "Found %/% left/right elements adjacent to a merge coordinate",
+           elements.leftElements.size(), elements.rightElements.size());
+
+    // Recurse over the dimensions for the MeshEntities of higher dimension.
+    computeMergePlanLevels(mesh, pairing, elements, merges, tag<1>{});
+
+    // Print some debugging information
+    logger(INFO, "Merge statistics:");
+    for (std::size_t i = 0; i < dimension; ++i) {
+        std::map<std::size_t, std::size_t> mergeSizes;
+        for (const auto& merge : merges[i]) {
+            mergeSizes[merge.size()]++;
+        }
+        logger(INFO, "\tmerges for entity dimension %", i);
+        for (const auto& sizePair : mergeSizes) {
+            logger(INFO, "\t\tsize %: %", sizePair.first, sizePair.second);
+        }
+    }
+    return MergePlan<dimension>(mesh, merges);
+}
+
+template <std::size_t dimension>
+template <std::size_t d>
+void MergePlan<dimension>::computeMergePlanLevels(
+    const Mesh<dimension>* mesh, const std::map<CoordId, CoordId>& pairing,
+    const MergeElements& elements,
+    std::array<std::vector<EntityMergeGroup>, dimension>& merges, tag<d>) {
+    static_assert(d < dimension, "Can't merge elements or higher dimension");
+
+    // Computing the merge plan for a single level done in two steps
+    //  1. Find all eligible left-entities. That is an Entity where there is an
+    //     Element such that all the nodes of the entity have a coordinate in
+    //     the left set of the coordinate pairing.
+    //     Map the coordinates to the corresponding right coordinates and store
+    //     the set of right coordinates with a reference to the left-entity.
+    //  2. Find all eligible right entities, and check if there is a matching
+    //     left-entity.
+
+    // Eligible left-entities, indexed by the mapped coordinates (i.e.
+    // coordinates in the right set).
+    std::map<std::set<CoordId>, EntityGId> leftCandidates;
+
+    for (const EntityGId& leftElementId : elements.leftElements) {
```

Comment:
I find this hard to read and understand, maybe more functions help?

Reply:
I've split of the left into a function. The whole loop over elements and then over boundary MeshEntities has also been abstracted away. So hopefully that makes things a bit easier.